cpp/ql/src/Critical/UseAfterFree.ql

```diff
@@ -11,56 +11,26 @@
  */
 
 import cpp
-import semmle.code.cpp.controlflow.StackVariableReachability
-
-/** `e` is an expression that frees the memory pointed to by `v`. */
-predicate isFreeExpr(Expr e, StackVariable v) {
-  exists(VariableAccess va | va.getTarget() = v |
-    exists(FunctionCall fc | fc = e |
-      fc.getTarget().hasGlobalOrStdName("free") and
-      va = fc.getArgument(0)
-    )
-    or
-    e.(DeleteExpr).getExpr() = va
-    or
-    e.(DeleteArrayExpr).getExpr() = va
-  )
-}
-
-/** `e` is an expression that (may) dereference `v`. */
-predicate isDerefExpr(Expr e, StackVariable v) {
-  v.getAnAccess() = e and dereferenced(e)
-  or
-  isDerefByCallExpr(_, _, e, v)
+import semmle.code.cpp.dataflow.new.DataFlow
+import FlowAfterFree as Flow
+
+module UseAfterFreeFlow = Flow::FlowFromFree<useExpr/2>;
+
+predicate useExpr(DataFlow::Node dfe, Expr e) {
+  e = dfe.asExpr() and
+  (
+    e = any(ReturnStmt rs).getExpr() or
+    e = any(FunctionCall fc).getAChild() or
+    e = any(PointerDereferenceExpr pde).getOperand() or
+    e = any(PointerFieldAccess pfa).getQualifier() or
+    e = any(ArrayExpr ae).getAChild()
```

Comment:
I hope this is sufficient? No need to also include the array offset in the set of sinks, I think.
```suggestion
    e = any(ArrayExpr ae).getArrayBase()
```

Reply:
Done